spec/unit/util/character_encoding_spec.rb

```diff
@@ -1,88 +1,118 @@
 #! /usr/bin/env ruby
 require 'spec_helper'
 require 'puppet/util/character_encoding'
+require 'puppet_spec/character_encoding'
 
 describe Puppet::Util::CharacterEncoding do
   describe "::convert_to_utf_8!" do
     context "when passed a string that is already UTF-8" do
       context "with valid encoding" do
-        it "should return the string unaltered" do
+        it "should not modify the string" do
           utf8_string = "\u06FF\u2603"
-          expect(Puppet::Util::CharacterEncoding.convert_to_utf_8!(utf8_string)).to eq(utf8_string)
+          Puppet::Util::CharacterEncoding.convert_to_utf_8!(utf8_string)
+          expect(utf8_string).to eq("\u06FF\u2603")
         end
       end
 
       context "with invalid encoding" do
         let(:invalid_utf8_string) { "\xfd\xf1".force_encoding(Encoding::UTF_8) }
 
         it "should issue a debug message" do
-          Puppet.expects(:debug).with(regexp_matches(/not valid UTF-8/))
+          Puppet.expects(:debug).with(regexp_matches(/encoding is invalid/))
           Puppet::Util::CharacterEncoding.convert_to_utf_8!(invalid_utf8_string)
         end
 
-        it "should return nil" do
-          expect(Puppet::Util::CharacterEncoding.convert_to_utf_8!(invalid_utf8_string)).to be_nil
+        it "should not modify the string" do
+          Puppet::Util::CharacterEncoding.convert_to_utf_8!(invalid_utf8_string)
+          expect(invalid_utf8_string).to eq("\xfd\xf1".force_encoding(Encoding::UTF_8))
         end
       end
     end
 
     context "when passed a string not in UTF-8 encoding" do
-      context "the bytes of which represent valid UTF-8" do
-        # I think this effectively what the ruby Etc module is doing when it
-        # returns strings read in from /etc/passwd and /etc/group
-        let(:iso_8859_1_string) { [225, 154, 160].pack('C*').force_encoding(Encoding::ISO_8859_1) }
-        let(:result) { Puppet::Util::CharacterEncoding.convert_to_utf_8!(iso_8859_1_string) }
-
-        it "should set external encoding to UTF-8" do
-          expect(result.encoding).to eq(Encoding::UTF_8)
-        end
+      it "should be able to convert BINARY to UTF-8 by labeling as Encoding.default_external" do
+        # そ - HIRAGANA LETTER SO
+        # In Windows_31J: \x82 \xbb - 130 187
+        # In Unicode: \u305d - \xe3 \x81 \x9d - 227 129 157
+
+        # When received as BINARY are not transcodable, but by "guessing"
+        # Encoding.default_external can transcode to UTF-8
+        win_31j = [130, 187].pack('C*')
```

Comment:
We should probably introduce a failure spec here that shows an incoming BINARY byte sequence that is *not* valid in `Encoding.default_external` ... i.e. it should raise an exception of `InvalidByteSequenceError`